src/charm.py

```diff
@@ -90,12 +178,100 @@ def _on_get_credentials_action(self, event: ActionEvent) -> None:
             result["postgresql"] = list(postgresql.values())[0]
         if mongodb:
             result["mongodb"] = list(mongodb.values())[0]
+        if kafka:
+            result["kafka"] = list(kafka.values())[0]
         event.set_results(result)
 
     def _on_database_created(self, event: DatabaseCreatedEvent) -> None:
         """Event triggered when a database was created for this application."""
         logger.info(f"database credentials are received: {event.username}")
-        self.unit.status = ActiveStatus(f"received {self.database} credentials")
+        self.unit.status = ActiveStatus(f"Received credentials for database: {self.database}.")
+
+    def _on_topic_created(self, event: TopicCreatedEvent) -> None:
+        """Event triggered when a topic was created for this application."""
+        logger.info(f"Topic credentials are received: {event.username}")
+        self.unit.status = ActiveStatus(f"Received credentials for topic: {self.topic}")
+
+    @property
+    def database(self):
+        """Return the configured database name."""
+        return self.get_secret("app", "database") if self.get_secret("app", "database") else ""
+
+    @property
+    def topic(self):
+        """Return the configured topic name."""
+        return self.get_secret("app", "topic") if self.get_secret("app", "topic") else ""
+
+    @property
+    def extra_user_roles(self):
+        """Return the configured user-extra-roles parameter."""
+        return (
+            self.get_secret("app", "extra-user-roles")
+            if self.get_secret("app", "extra-user-roles")
+            else ""
+        )
+
+    @property
+    def is_database_related(self):
+        """Return if a relation with database is present."""
+        if self.mysql.relations:
+            return True
+        if self.postgresql.relations:
+            return True
+        if self.mongodb.relations:
+            return True
+        # no reltation with databases
+        return False
+
+    @property
+    def is_kafka_related(self):
+        """Return if a relation with kafka is present."""
+        if self.kafka.relations:
+            return True
+        # no reltation with kafka
+        return False
+
+    @property
+    def app_peer_data(self) -> Dict:
+        """Application peer relation data object."""
+        relation = self.model.get_relation(PEER)
+        if not relation:
+            return {}
+
+        return relation.data[self.app]
+
+    @property
+    def unit_peer_data(self) -> Dict:
+        """Peer relation data object."""
+        relation = self.model.get_relation(PEER)
+        if relation is None:
+            return {}
+
+        return relation.data[self.unit]
+
+    def get_secret(self, scope: str, key: str) -> Optional[str]:
```

Comment:
We should really factor these methods out to a common class in a charm library or similar....

We are doing a lot of copy and pasted every time. 

But let's address this item properly. No need to follow up on this within this PR

Reply:
I don't understand your point. I mean, we could just have a MixinClass or a CharmBase-derived class in our charm or python libraries, with useful methods we generally use:

```python 

class DataPlatformCharm(CharmBase): #or mixin if we don't like inheritance

    def get_secret(...):
        # Implementation

    def set_secret(...):
        # Implementation

```

Why are we bounded to the OperatorFramework? I really don't see this going there or being rather hard to push these there. If - as dataplatform - we really like these methods, I'd factor them out in a lib (being charm or normal python). This sounds a lot better than copy and paste

Reply:
These `set-secret` `get-secrets` are already meant to be removed as soon as an equivalent method exists in Operator Framework, and we added them only because it would make refactoring easier when that time came around if we had only 2 method names to reference. I would avoid investing more time on a stop-gap.

Reply:
As said, no need to address this issue here. However, generally speaking, I can see value on having a process for implementing/experimenting things in data platform charm libraries (or packages) to be used within our projects first to then promote such features to the Operator Framework.